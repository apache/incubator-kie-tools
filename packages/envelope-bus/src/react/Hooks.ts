/*
 * Copyright 2020 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ApiDefinition, MessageBusClient, NotificationPropertyNames, SubscriptionCallback } from "../api";
import { useEffect } from "react";

export function useSubscription<Api extends ApiDefinition<Api>, M extends NotificationPropertyNames<Api>>(
  bus: MessageBusClient<Api>,
  method: M,
  callback: SubscriptionCallback<Api, M>
) {
  useEffect(() => {
    const subscription = bus.subscribe(method, callback);
    return () => {
      bus.unsubscribe(method, subscription);
    };
  }, [bus, method, callback]);
}

export function useSubscriptionOnce<Api extends ApiDefinition<Api>, M extends NotificationPropertyNames<Api>>(
  bus: MessageBusClient<Api>,
  method: M,
  callback: SubscriptionCallback<Api, M>
) {
  useEffect(() => {
    let unsubscribed = false;

    const subscription = bus.subscribe(method, (...args) => {
      callback(...args);
      unsubscribed = true;
      bus.unsubscribe(method, subscription);
    });

    return () => {
      if (!unsubscribed) {
        bus.unsubscribe(method, subscription);
      }
    };
  }, [bus, method, callback]);
}
