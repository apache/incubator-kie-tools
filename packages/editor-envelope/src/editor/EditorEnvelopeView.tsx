/*
 * Copyright 2020 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from "react";
import { Editor } from "@kogito-tooling/editor-api";
import { LoadingScreen } from "./LoadingScreen";
import "@patternfly/patternfly/base/patternfly-variables.css";
import "@patternfly/patternfly/patternfly-addons.scss";
import "@patternfly/patternfly/patternfly.scss";
import { KeyBindingsHelpOverlay } from "./KeyBindingsHelpOverlay";

interface Props {
  exposing: (self: EditorEnvelopeView) => void;
}

interface State {
  editor?: Editor;
  loading: boolean;
}

export class EditorEnvelopeView extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = { editor: undefined, loading: true };
    this.props.exposing(this);
  }

  public getEditor() {
    return this.state.editor;
  }

  public setEditor(editor: Editor) {
    return new Promise(res => this.setState({ editor: editor }, res));
  }

  public setLoadingFinished() {
    return new Promise(res => this.setState({ loading: false }, res));
  }

  public setLoading() {
    return this.setState({ loading: true });
  }

  public render() {
    return (
      <>
        {!this.state.loading && <KeyBindingsHelpOverlay />}
        <div id="loading-screen" style={{ zIndex: 100, position: "relative" }}>
          <LoadingScreen visible={this.state.loading} />
        </div>
        {this.state.editor && this.state.editor.af_isReact && this.state.editor.af_componentRoot()}
      </>
    );
  }
}
