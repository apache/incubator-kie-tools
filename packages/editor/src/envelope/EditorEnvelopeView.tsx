/*
 * Copyright 2020 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from "react";
import { Editor } from "../api";
import { LoadingScreen } from "./LoadingScreen";
import { KeyBindingsHelpOverlay } from "./KeyBindingsHelpOverlay";
import { useCallback, useImperativeHandle, useState } from "react";

interface Props {
  setLocale: React.Dispatch<string>;
  showKeyBindingsOverlay: boolean;
}

export interface EditorEnvelopeViewApi<E extends Editor> {
  getEditor: () => E | undefined;
  setEditor: (editor: E) => void;
  setLoading: () => void;
  setLoadingFinished: () => void;
  setLocale: (locale: string) => void;
}

export const EditorEnvelopeViewRef: React.ForwardRefRenderFunction<EditorEnvelopeViewApi<Editor>, Props> = (
  props: Props,
  forwardingRef
) => {
  const [editor, setEditor] = useState<Editor | undefined>(undefined);
  const [loading, setLoading] = useState(true);

  useImperativeHandle(
    forwardingRef,
    () => {
      return {
        getEditor: () => editor,
        setEditor: (newEditor) => setEditor(newEditor),
        setLoading: () => setLoading(true),
        setLoadingFinished: () => setLoading(false),
        setLocale: (locale) => props.setLocale(locale),
      };
    },
    [props, editor]
  );

  return (
    <>
      {!loading && props.showKeyBindingsOverlay && <KeyBindingsHelpOverlay />}
      <LoadingScreen loading={loading} />
      <div style={{ position: "absolute", width: "100vw", height: "100vh", top: "0", left: "0" }}>
        {editor && editor.af_isReact && editor.af_componentRoot()}
      </div>
    </>
  );
};

export const EditorEnvelopeView = React.forwardRef(EditorEnvelopeViewRef);
