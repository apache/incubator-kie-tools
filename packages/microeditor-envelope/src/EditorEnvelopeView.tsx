/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from "react";
import * as ReactDOM from "react-dom";
import * as AppFormer from "@kogito-tooling/core-api";
import { LoadingScreen } from "./LoadingScreen";

interface Props {
  exposing: (self: EditorEnvelopeView) => void;
  loadingScreenContainer: HTMLElement;
}

interface State {
  editor?: AppFormer.Editor;
  loading: boolean;
}

export class EditorEnvelopeView extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = { editor: undefined, loading: true };
    this.props.exposing(this);
  }

  public getEditor() {
    return this.state.editor;
  }

  public setEditor(editor: AppFormer.Editor) {
    return new Promise(res => this.setState({ editor: editor }, res));
  }

  public setLoadingFinished() {
    return new Promise(res => this.setState({ loading: false }, res));
  }

  private LoadingScreenPortal() {
    return ReactDOM.createPortal(<LoadingScreen visible={this.state.loading} />, this.props.loadingScreenContainer!);
  }

  public render() {
    return (
      <>
        {this.LoadingScreenPortal()}
        {this.state.editor && this.state.editor.af_isReact && this.state.editor.af_componentRoot()}
      </>
    );
  }
}
