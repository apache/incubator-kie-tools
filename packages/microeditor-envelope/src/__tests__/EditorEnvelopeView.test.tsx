/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from "react";
import { shallow } from "enzyme";
import { EditorEnvelopeView } from "../EditorEnvelopeView";
import { DummyEditor } from "./DummyEditor";

let loadingScreenContainer: HTMLElement;
beforeEach(() => (loadingScreenContainer = document.body.appendChild(document.createElement("div"))));
afterEach(() => loadingScreenContainer.remove());

function renderEditorEnvelopeView(): [EditorEnvelopeView, ReturnType<typeof shallow>] {
  let view: EditorEnvelopeView;
  const render = shallow(
    <EditorEnvelopeView exposing={self => (view = self)} loadingScreenContainer={loadingScreenContainer} />
  );
  return [view!, render];
}

describe("EditorEnvelopeView", () => {
  test("first open", () => {
    const [_, render] = renderEditorEnvelopeView();
    expect(render).toMatchSnapshot();
  });

  test("after loading stops", async () => {
    const [view, render] = renderEditorEnvelopeView();
    await view.setLoadingFinished();
    expect(render).toMatchSnapshot();
  });

  test("after loading stops and editor is set", async () => {
    const [view, render] = renderEditorEnvelopeView();
    await view.setLoadingFinished();
    await view.setEditor(new DummyEditor());
    expect(render).toMatchSnapshot();
  });
});
