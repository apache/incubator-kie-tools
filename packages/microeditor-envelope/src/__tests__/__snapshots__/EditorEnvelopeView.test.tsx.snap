// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorEnvelopeView after loading stops 1`] = `
<Fragment>
  <Portal
    containerInfo={<div />}
  >
    <LoadingScreen
      visible={false}
    />
  </Portal>
</Fragment>
`;

exports[`EditorEnvelopeView after loading stops and editor is set 1`] = `
<Fragment>
  <Portal
    containerInfo={<div />}
  >
    <LoadingScreen
      visible={false}
    />
  </Portal>
  <DummyEditorComponent
    exposing={[Function]}
  />
</Fragment>
`;

exports[`EditorEnvelopeView first open 1`] = `
<Fragment>
  <Portal
    containerInfo={<div />}
  >
    <LoadingScreen
      visible={true}
    />
  </Portal>
</Fragment>
`;
